File: blend_stats.php
<?php 
/*
This file is part of BlendStats.

BlendStats is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

BlendStats is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with BlendStats.  If not, see <http://www.gnu.org/licenses/>.
*/


if ( ! function_exists('debug') ) {
	/**
	 * Utility debugging function, defined only if no global version exists
	 * @param  mixed $var Anything that print_r() can print
	 */
	function debug( $var = null ) {
		echo '<pre>';
		print_r( $var );
		echo '</pre>';
	}
}

/**
 * A class to be used in your PHP application for access to blend file stats
 * Usage:
 * 		$bst = new BlendStats( $path_to_blend_file );
 * 		$stats = $bst->get_stats();
 * 		// do operations on $stats
 */
class BlendStats {

	public $start_marker = "---STATS---BEGIN---";

	public $end_marker = "---STATS---END---";

	public $blend_path = null;

	public $blend_dir = null;

	public $blender_bin = null;

	public $stats = null;

	/**
	 * Constructor for this class
	 * you should pass the blend file path and the Blender binary call as it 
	 * will be called
	 * @param [type] $file_path   full path of the blend file you want to read
	 * @param [type] $blender_bin Blender binary, use if you don't have Blender in $PATH
	 */
	public function __construct( $fpath = null, $bin = 'blender' ) {
		if ( ! defined('DS') ) {
			define('DS', DIRECTORY_SEPARATOR);
		}
		if ( file_exists($fpath) ) {
			$this->blend_path = $fpath;
			$this->blender_bin = $bin;
		}
		$this->script_path = dirname(__FILE__) . DS . 'blend_stats.py';
	}

	/**
	 * Reads file from disk and return full uotput from Blender
	 * @param  string $path the absolute path to the blend file
	 * @return string       the full Blender output on runtime
	 */
	public function get_blender_output() {
		$output = "";
		$output = shell_exec($this->blender_bin .' -noaudio -d -y -Y -b '. $this->blend_path .' --python '. $this->script_path .' --verbose 2 -- ' . $this->blend_dir);
		return $output;
	}

	/**
	 * [get_stats description]
	 * @return [type] [description]
	 */
	public function get_stats() {
		if ( ! $this->stats ) {
			$this->load_stats();
		}
		return $this->stats;
	}

	/**
	 * Cleans the raw output from get_blender_output and returns a PHP object 
	 * containing all the stats for direct access
	 * @param  [type] $path [description]
	 * @return stdClass     Standard object with stats
	 */
	public function load_stats() {
		if ( ! $this->blend_path ) {
			return null;
		}
		$raw_output = $this->get_blender_output( $this->blend_path );
		$json_string = $this->isolate_json( $raw_output );
		$this->stats = json_decode( $json_string );
		return $this->stats;
	}

	/**
	 * Takes the raw output and returns the json string we want
	 * @param  string $raw Raw output from Blender
	 * @return string      the isolated json string in the Blender output
	 */
	private function isolate_json( $raw = null ) {
		$start_pos = strpos($raw, $this->start_marker);
		$end_pos = strpos($raw, $this->end_marker);
		if ( false === $start_pos || false === $end_pos ) {
			return null;
		}
		$start_pos += strlen( $this->start_marker );
		
		$json_string = substr($raw, $start_pos, $end_pos - strlen($raw) );

		return trim( $json_string );
	}


	public function test() {
		$test_file = dirname(__FILE__) . DS . "test/test.blend";
		$this->get_stats( $test_file );
		debug( $this->stats );
	}

}

File: blend_stats.py
"""
This file is part of BlendStats.

BlendStats is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

BlendStats is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with BlendStats.  If not, see <http://www.gnu.org/licenses/>.
"""
# <pep8-80 compliant>

############
## CONFIG ##
###################################################################

# Should get absolute basepath from argument params[0]
base_path = "D:\\Webserver\\xampp\\htdocs\\blend_stats\\"

###################################################################

import bpy
import json
import os
import sys

context = bpy.context

# Workaround - ensure object mode, even if scene is actually
# in an invalid state (context manipulated by evil script)
def toggle_object_mode():

    limit = 3  # 3 iterations at most
    while ('OBJECT' != context.mode and limit > 0 and
           bpy.ops.object.editmode_toggle.poll()):

        # mode_set(mode='OBJECT') broken in background mode up to r57205
        bpy.ops.object.editmode_toggle()
        limit -= 1

    if context.mode != 'OBJECT':
        # Should only occur if above poll failed (ever?)
        print("Error: mode still not 'OBJECT'!")
        print("Scene statistics output is going to be incorrect.")


def isabs(filepath):
    #return os.path.isabs(filepath)
    #return not (filepath.startswith("//") or ".." in filepath)
    return not filepath.startswith("//")


def is_valid_path(filepath, basepath):
    # print("basepath", basepath)
    # print("filepath", filepath)
    if not filepath.startswith("//"):
        print("no leading //")
        return False
    realpath = os.path.realpath(bpy.path.abspath(filepath))
    # print("realpath", realpath)
    return (realpath.startswith(basepath) and
            os.path.exists(realpath) and
            os.path.isfile(realpath))

# Vars for holding results
output = {}
scenes = []


for scene in bpy.data.scenes:

    # Setting active scene resets object mode in regular Blender use.
    # Unfortunately, it doesn't switch if running in background mode.
    # However, Scene.statistics() still requires the scene being active!
    context.screen.scene = scene

    # Ensure 'OBJECT' mode
    toggle_object_mode()

    verts = sum(len(ob.data.vertices) \
        for ob in scene.objects if ob.type == 'MESH')

    polygons = sum(len(ob.data.polygons) \
        for ob in scene.objects if ob.type == 'MESH')

    render_engine = scene.render.engine

    objects_mesh = len([ob for ob in scene.objects if ob.type == 'MESH'])

    # Get scene statistics from built-in function (object mode required)
    # It includes modifiers (final meshes)!
    builtin_stats = scene.statistics()

    # Extract relevant information
    for segment in builtin_stats.split(" | "):
        segment_split = segment.split(":")
        if segment_split[0] == "Verts":
            verts_modified = int(segment_split[1])
        elif segment_split[0] == "Faces":
            polygons_modified = int(segment_split[1])
        elif segment_split[0] == "Tris":
            triangles = int(segment_split[1])

    this_scene = {
        'name': scene.name,
        'verts': verts,
        'verts_modified': verts_modified,
        'polygons': polygons,
        'polygons_modified': polygons_modified,
        'triangles': triangles,
        'render_engine': render_engine,
        'objects_mesh': objects_mesh,
    }
    scenes.append(this_scene)

output['scenes'] = scenes


# Get script parameters:
# all list items after the last occurence of "--"
print()
print(sys.argv)
print()

try:
    args = list(reversed(sys.argv))
    idx = args.index("--")

except ValueError:
    params = []

else:
    params = args[:idx][::-1]

print("Script params:", params)


# Check for broken image references
bad_images = []

for image in bpy.data.images:

    if image.name in bad_images:
        continue

    if ((image.users > 0 # image datablock in use (ignores use_fake_user)
         and image.packed_file is None # image not stored inside the .blend
         and image.filepath # filepath string is not empty
         # valid means: blender-style relative path, inside base path,
         # path actually exists and path references a file. Checks for invalid.
         and not is_valid_path(image.filepath, base_path)
        )
        or image.library is not None): # is linked from another .blend

        bad_images.append(image.name)

output['bad_images'] = bad_images

""" pretty useless actually, as it's not about the .blend file
output['version'] = {
    "char": bpy.app.version_char,
    "cycle": bpy.app.version_cycle,
    "number": bpy.app.version,
    "revision": bpy.app.build_revision.decode(),

}
"""

# Dump gathered information to command line
# JSON module safely encodes UTF8/16 chars, as well as escape sequences
print("\n---STATS---BEGIN---")
print(json.dumps(output, indent=4, sort_keys=True))
print("---STATS---END---")
